### PR TITLE
refactor: search screen to use relay hooks

### PR DESCRIPTION
**Pull request by gkartalis**
The type of this PR is: **Enhancement**

### Description

Search screen refactor to take advantage of the new relay hooks api. 
Linking here [Relay Hooks RFC](https://github.com/artsy/eigen/issues/6080)

### PR Checklist (tick all before merging)

- [x] I have included screenshots or videos to illustrate my changes, or I have not changed anything that impacts the UI.
- [x] I have tested my changes on **iOS** and **Android**.
- [x] I have added tests/stories for my changes, or my changes don't require testing/stories, or I have included a link to a separate Jira ticket covering the tests.
- [x] I have added a feature flag, or my changes don't require a feature flag. ([How do I add one?](https://github.com/artsy/eigen/blob/main/docs/developing_a_feature.md))
- [x] I have documented any follow-up work that this PR will require, or it does not require any.
- [x] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/main/docs/adding_state_migrations.md))
- [x] I have added a changelog entry below or my changes do not require one.

### To the reviewers 👀

- [ ] I would like at least one of the reviewers to run this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

- refactor search screen to use relay hooks - gkartalis

</details>

### CODE REVIEW

File: src/lib/Scenes/Search/components/placeholders/SearchPlaceholder.tsx [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
+  ProvidePlaceholderContext,
+} from "lib/utils/placeholders"
 import { times } from "lodash"
 import { Box, Flex, Spacer } from "palette"
 import React from "react"
 
 export const SearchPlaceholder: React.FC = () => {
   return (
-    <>
+    <ProvidePlaceholderContext>
```
Comment (gkartalis) [inline src/lib/Scenes/Search/components/placeholders/SearchPlaceholder.tsx:13]: Moved the `ProvidePlaceholderContext` cause I noticed that we use it to wrap the `SearchPlaceholder` everywhere we use it

File: src/lib/Scenes/Search/Search.tsx [hunk 2]
```diff
@@ -259,59 +268,17 @@ export const Search: React.FC<SearchProps> = (props) => {
...
-      environment={defaultEnvironment}
-      query={SearchScreenQuery}
-      render={({ props, error }) => {
-        if (error) {
-          if (__DEV__) {
-            console.error(error)
-          } else {
-            captureMessage(error.stack!)
-          }
-        }
```
Comment (gkartalis) [inline src/lib/Scenes/Search/Search.tsx:309]: 🤔  how can we capture the errors now with the hooks api? Should we make a universal `<ErrorBoundary renderError={Error}>` component to wrap components and track all screen errors?

cc @olerichter00

Comment (olerichter00) [reply src/lib/Scenes/Search/Search.tsx:309]: We already wrap our app with an `ErrorBoundary` [here](https://github.com/artsy/eigen/blob/d4cfd3fb0f045e5fda44a8387120a701358a03f6/src/lib/AppProviders.tsx#L21). In case you need special error handling, you can add one to `SearchScreen`.

File: src/lib/Scenes/Search/components/placeholders/SearchPlaceholder.tsx (hunk 1 above)
Comment (olerichter00) [reply src/lib/Scenes/Search/components/placeholders/SearchPlaceholder.tsx:13]: 👍

**Review by olerichter00 — APPROVED** [review]
Great work 🌟 

I've just added a question about using the placeholder in `Search.tsx`.

File: src/lib/Scenes/Search/RefetchWhenApiKeyExpired.tsx [hunk 3]
```diff
@@ -1,33 +1,21 @@
...
-      },
-      { force: true }
-    )
-  }
+  const { error, refetch } = props
 
   useEffect(() => {
     if (isAlgoliaApiKeyExpiredError(error)) {
-      refetch()
+      refetch({})
```
Comment (dimatretyak) [inline src/lib/Scenes/Search/RefetchWhenApiKeyExpired.tsx:18]: question: maybe you should specify `fetchPolicy: "network-only"` for `refetch`? Otherwise, we will reuse locally cached data ([documentation](https://relay.dev/docs/v12.0.0/guided-tour/reusing-cached-data/fetch-policies/))

image

**Comment by dimatretyak** [comment]
🚀 Looks great. Thank you for these changes. I just noticed one small point: previously, the token refetch was "silent", but now loading placeholder is displayed every time.

### Prev implementation
https://user-images.githubusercontent.com/3513494/153042326-1292fbcb-b03a-4cc5-a23a-56819522686a.mp4

### New implementation
https://user-images.githubusercontent.com/3513494/153042370-f2d3a277-a710-42cb-a2c6-cbe3a87794d2.mp4

I see only one edge case and bad UX: when the user makes request to algolia (for example, scrolls the list down or switches pills) and token has expired.

File: src/lib/Scenes/Search/Search.tests.tsx [hunk 4]
```diff
@@ -524,33 +650,45 @@ describe("Search Screen", () => {
...
+    const searchInput = getByPlaceholderText("Search artists, artworks, galleries, etc")
+    act(() => fireEvent(searchInput, "changeText", "text"))
 
-    mockEnvironmentPayload(mockEnvironment, {
-      SearchableConnection: () => ({
-        edges: [
-          {
-            node: {
-              displayLabel: "Banksy",
+    mockEnvironment.mock.resolveMostRecentOperation((operation) =>
```
Comment (gkartalis) [inline src/lib/Scenes/Search/Search.tests.tsx:674]: The mostRecentOperation here was the initial problem because it couldn't be found in the queue since the `Search` component relies on relay-hooks and the other one on the containers. As a result I had to import the default environment (a separate one mocked in setupJest) in order to `resolveMostRecentOperation` successfully.

File: src/lib/Scenes/Search/Search.tests.tsx [hunk 5]
```diff
@@ -49,28 +50,48 @@ jest.mock("lodash", () => ({
...
+        <SearchScreen />
+      </RelayEnvironmentProvider>
+    )
   }
 
   it("should render a text input with placeholder", async () => {
-    const { getByPlaceholderText, getByText, queryByText } = renderWithWrappersTL(<TestRenderer />)
+    const { getByPlaceholderText, getByText, queryByText } = await renderWithWrappersTL(
+      <TestRenderer />
+    )
```
Comment (gkartalis) [inline src/lib/Scenes/Search/Search.tests.tsx:79]: This was the only way I came up with having the SAME relay test environment in order to make the [final test pass](https://github.com/artsy/eigen/pull/6164/files#r802988822). 

Another way that I found while I was experimenting is to wrap the `TestRenderer` with two `RelayEnvironmentProviders`, one that had the mockEnvironment (created once before each test) and the parent one had the default environment.

I wasn't a huge fan of this approach so I went with the changes that you see here. 

cc @pvinis @araujobarret @MounirDhahri @olerichter00 let me know what you think.

File: src/lib/Scenes/Search/Search.tests.tsx (hunk 4 above)
Comment (olerichter00) [reply src/lib/Scenes/Search/Search.tests.tsx:674]: I've just had the same issue.

File: src/lib/Scenes/Search/Search.tests.tsx (hunk 5 above)
Comment (olerichter00) [reply src/lib/Scenes/Search/Search.tests.tsx:79]: I like this approach. Can we bring `reset` to the mock in [setupJest.ts](https://github.com/artsy/eigen/blob/7430c0935b27e169f1a7a50450d9b6b4b3180fe5/src/setupJest.ts#L42)?

Comment (MounirDhahri) [reply src/lib/Scenes/Search/Search.tests.tsx:79]: Good approach

File: src/lib/Scenes/Search/Search.tests.tsx (hunk 4 above)
Comment (gkartalis) [reply src/lib/Scenes/Search/Search.tests.tsx:667]: Let me know if you need any help with this 🤘

File: src/lib/Scenes/Search/Search.tests.tsx (hunk 5 above)
Comment (araujobarret) [reply src/lib/Scenes/Search/Search.tests.tsx]: We tried a few things and didn't work out, this mixed hooks, hocs and defaultEnvironment is annoying to deal with.
It's okay ✅  but is a bit confusing and hard to read for people that don't have the context of the situation, unfortunately, I have no other solution so far

File: src/lib/Scenes/Search/Search.tests.tsx (hunk 4 above)
Comment (olerichter00) [reply src/lib/Scenes/Search/Search.tests.tsx:667]: Using `useRelayEnvironment` instead of importing `defaultEnvironment` might fix the problem as well ([documentation](https://relay.dev/docs/guided-tour/rendering/environment/#accessing-the-relay-environment)).

File: src/lib/Scenes/Search/Search.tsx [hunk 6]
```diff
@@ -122,11 +135,7 @@ export const Search: React.FC<SearchProps> = (props) => {
   }, [indices, indicesInfo])
 
   if (!searchClient || !searchInsightsConfigured) {
-    return (
-      <ProvidePlaceholderContext>
-        <SearchPlaceholder />
-      </ProvidePlaceholderContext>
-    )
+    return <SearchPlaceholder />
```
[the remark this answers is not in the document]
Comment (gkartalis) [reply src/lib/Scenes/Search/Search.tsx:138]: unfortunately not, this has to do with the useSearchInsights hook that we have and with the initialization of algolia client so cant be caught for now

File: src/lib/Scenes/Search/Search.tests.tsx (hunk 5 above)
Comment (olerichter00) [reply src/lib/Scenes/Search/Search.tests.tsx]: Using `useRelayEnvironment` instead of importing `defaultEnvironment` might be cleaner and solve these issues.

Comment (gkartalis) [reply src/lib/Scenes/Search/Search.tests.tsx]: not sure I understand 100% how this can be done. Where exactly do you suggest using the `useRelayEnvironment` instead of `defaultEnvironment`?

Comment (araujobarret) [reply src/lib/Scenes/Search/Search.tests.tsx]: I guess @olerichter00 meant in the components instead of importing the singleton we have, just getting from the context with the hook `useRelayEnvironment`, this solution would save us from dealing with `defaultEnvironment` all the time

**Review by dimatretyak — APPROVED** [review]

**Review by olerichter00 — APPROVED** [review]
🔍  🪟   🚀  🚀  🚀

File: .secrets.baseline [hunk 7]
```diff
@@ -66,10 +66,6 @@
     {
       "path": "detect_secrets.filters.allowlist.is_line_allowlisted"
     },
-    {
-      "path": "detect_secrets.filters.common.is_baseline_file",
-      "filename": ".secrets.baseline"
-    },
```
Comment (pvinis) [inline .secrets.baseline:72]: um.... this shouldnt go away 🤔 

did you have to run the generate secrets script?

Comment (gkartalis) [reply .secrets.baseline:72]: yeah I did run it because I added some lines to setup jest, the pre push hook prevented me from pushing due to this and I generated a new one and commited. 🤔 Is it something that I didn't do correctly?
